adapters/insticator/insticator.go

```diff
@@ -0,0 +1,335 @@
+		}
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request openrtb2.BidRequest, impList []openrtb2.Imp) (*adapters.RequestData, error) {
+	request.Imp = impList
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+			headers.Add("IP", request.Device.IP)
+		}
+	}
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(responseData) {
+		return nil, nil
+	}
+
+	if err := adapters.CheckResponseStatusCodeForErrors(responseData); err != nil {
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	if response.Cur != "" {
+		bidResponse.Currency = response.Cur
+	}
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bid := &seatBid.Bid[i]
+			bidType := getMediaTypeForBid(bid)
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+	return bidResponse, nil
+}
+
+func makeImps(imp openrtb2.Imp) (openrtb2.Imp, string, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var insticatorExt openrtb_ext.ExtImpInsticator
+	if err := json.Unmarshal(bidderExt.Bidder, &insticatorExt); err != nil {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	// check if the adUnitId is not empty
+	if insticatorExt.AdUnitId == "" {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: "Missing adUnitId",
+		}
+	}
+
+	// check if the publisherId is not empty
+	if insticatorExt.PublisherId == "" {
+		return openrtb2.Imp{}, insticatorExt.AdUnitId, &errortypes.BadInput{
+			Message: "Missing publisherId",
+		}
+	}
+
+	// Directly construct the impExt
+	impExt := ext{
+		Insticator: impInsticatorExt{
+			AdUnitId:    insticatorExt.AdUnitId,
+			PublisherId: insticatorExt.PublisherId,
+		},
+	}
+
+	impExtJSON, err := json.Marshal(impExt)
```

Comment:
Please use our jsonutil helper as a drop in replacement for json. This is a new policy as of PBS 3.0 for performance.